pkg/genericapiserver/openapi/openapi.go

```diff
@@ -0,0 +1,485 @@
+	"net/http"
+	"reflect"
+	"strconv"
+	"strings"
+
+	"github.com/emicklei/go-restful"
+	"github.com/emicklei/go-restful/swagger"
+	"github.com/go-openapi/spec"
+)
+
+// Config is set of configuration for openAPI spec generation.
+type Config struct {
+	// SwaggerConfig is set of configuration for go-restful swagger spec generation. Currently
+	// openAPI implementation depends on go-restful to generate models.
+	SwaggerConfig *swagger.Config
+	// Info is general information about the API.
+	Info          *spec.Info
+	// FixOperation is a customization function that will be called on each operation.
+	// This function is optional.
+	FixOperation func(path string, op *spec.Operation, repeated bool)
+	// FixPath is a customization function that will be called on all generated path strings.
+	// This function is optional.
+	FixPath      func(path string) string
+	// FixResponses is a customization function that is called on all responses.
+	// This function is optional.
+	FixResponses func(responses *spec.Responses)
+	// List of webservice's path prefixes to ignore
+	IgnorePrefixes []string
+}
+
+type openAPI struct {
+	config       *Config
+	swagger      *spec.Swagger
+}
+
+// RegisterOpenAPIService registers a handler for /swagger.json and provides standard
+// swagger 2.0 (aka openAPI) specification.
+func RegisterOpenAPIService(config *Config, containers *restful.Container) (err error) {
+	o := openAPI{
+		config: config,
+	}
+	err = o.buildSwaggerSpec()
+	if err != nil {
+		return err
+	}
+	containers.ServeMux.HandleFunc("/swagger.json", func(w http.ResponseWriter, r *http.Request) {
+		resp := restful.NewResponse(w)
+		if r.URL.Path != "/swagger.json" {
+			resp.WriteErrorString(http.StatusNotFound, "Path not found!")
+		}
+		resp.WriteAsJson(o.swagger)
+	})
+	return nil
+}
+
+func (o *openAPI) buildSwaggerSpec() (err error) {
+	if o.swagger != nil {
+		return fmt.Errorf("Swagger spec is already built. Duplicate call to buildSwaggerSpec is not allowed.")
+	}
+	definitions, err := o.buildModels()
+	if err != nil {
+		return err
+	}
+	paths, err := o.buildPaths()
+	if err != nil {
+		return err
+	}
+	o.swagger = &spec.Swagger{
+		SwaggerProps: spec.SwaggerProps{
+			Swagger:     "2.0",
+			Definitions: definitions,
+			Paths:       &paths,
+			Info:        o.config.Info,
+		},
+	}
+	return nil
+}
+
+func (o *openAPI) buildModels() (definitions spec.Definitions, err error) {
+	definitions = spec.Definitions{}
+	for _, decl := range swagger.NewSwaggerBuilder(*o.config.SwaggerConfig).ProduceAllDeclarations() {
+		for _, oldNamedModel := range decl.Models.List {
+			_, ok := definitions[oldNamedModel.Name]
+			if ok {
+				// TODO(mbohlool): decide what to do with repeated models
+				continue
+			}
+			definitions[oldNamedModel.Name], err = buildModel(oldNamedModel.Model)
+			if err != nil {
+				return spec.Definitions{}, err
+			}
+		}
+	}
+	return definitions, nil
+}
+
+func buildModel(oldModel swagger.Model) (ret spec.Schema, err error) {
```

Comment:
I avoided swaggerModel because the new model is still technically swagger (2.0). I would however like this naming if I add a comment to the top of file explaining swagger here means swagger 1.2 and not swagger 2.0 (that is first version OpenAPI).